python/lsst/ts/genericcamera/genericcameracsc.py

```diff
@@ -705,7 +626,164 @@ async def run_auto_exposure_loop(self, min_exp_time, max_exp_time, configuration
         self.isAutoExposure = False
         self.log.info("autoExposure_loop - End")
 
-    async def establish_exposure_background(self, exposure):
+    async def run_auto_exposure(self, min_exp_time, max_exp_time, configuration):
+        """Take auto exposures until `runAutoExposureTask` is False.
+
+        Parameters
+        ----------
+        min_exp_time: `float`
+            The minimum exposure time to use.
+        max_exp_time: `float`
+            The maximum exposure time to use.
+        configuration: `dict`
+            A dict containing additional configuration parameters.
+        """
+        while self.runAutoExposureTask:
+            timestamp = time.time()
+            image_name = self.fileNameFormat.format(
+                timestamp=int(timestamp),
+                index=configuration["image_index"],
+                total=configuration["images_in_sequence"],
+            )
+
+            self.evt_startTakeImage.put()
+            await self.camera.startTakeImage(
+                configuration["exposure_time"],
+                configuration["shutter"],
+                configuration["sensors"],
+                configuration["keyValueMap"],
+                configuration["obsNote"],
+            )
+            exposure = await self.determine_auto_exposure_time_and_get_exposure(
+                min_exp_time, max_exp_time, configuration, timestamp, image_name
+            )
+
+            if exposure:
+                exposure.save(os.path.join(self.directory, image_name + ".fits"))
+
+            await self.camera.endTakeImage()
+            self.evt_endTakeImage.put()
+
+            # Schedule the next image such that it starts
```

Comment:
Here is an alternative idea. Create a task with the sleep before you call `self.determine_auto_exposure_time_and_get_exposure`, then here you can simply wait for the task to complete. If the task is already done, it is going to move forward, if not, it will wait for the task to complete:

```
task_timer = asyncio.create_task(asyncio.sleep(self.config.autoExposureInterval))
exposure = await self.determine_auto_exposure_time_and_get_exposure(
                 min_exp_time, max_exp_time, configuration, timestamp, image_name
             )

...

await task_timer

```

Reply:
THANK YOU!!! That is an awesome design pattern for scheduling recurring tasks and I have been looking for this for a long time now. Done.